File: gravitee-apim-definition/gravitee-apim-definition-model/src/main/java/io/gravitee/definition/model/v4/sharedpolicygroup/SharedPolicyGroup.java
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.definition.model.v4.sharedpolicygroup;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.gravitee.definition.model.Plugin;
import io.gravitee.definition.model.v4.flow.step.Step;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import java.io.Serializable;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

/**
 * @author Yann TAVERNIER (yann.tavernier at graviteesource.com)
 * @author GraviteeSource Team
 */
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@ToString
@EqualsAndHashCode
@Builder(toBuilder = true)
public class SharedPolicyGroup implements Serializable {

    @JsonProperty(required = true)
    @NotBlank
    private String id;

    @JsonProperty(required = true)
    @NotBlank
    private String environmentId;

    @JsonProperty(required = true)
    @NotBlank
    private String name;

    @JsonProperty(required = false)
    @NotBlank
    private String version;

    @JsonProperty(required = true)
    @NotBlank
    private Phase phase;

    @Valid
    private List<Step> policies;

    @JsonIgnore
    public List<Plugin> getPlugins() {
        return this.policies == null
            ? List.of()
            : this.policies.stream().filter(Step::isEnabled).map(Step::getPlugins).flatMap(List::stream).toList();
    }

    public enum Phase {
        REQUEST,
        RESPONSE,
        MESSAGE_REQUEST,
        MESSAGE_RESPONSE,
    }
}

File: gravitee-apim-console-webui/src/entities/management-api-v2/sharedPolicyGroup/SharedPolicyGroup.ts
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiType, StepV4 } from '../api';
import { ExecutionPhase } from '../plugin';

export interface SharedPolicyGroup {
  id: string;
  crossId: string;
  name: string;
  description?: string;
  prerequisiteMessage?: string;
  lifecycleState?: 'DEPLOYED' | 'UNDEPLOYED' | 'PENDING';
  version?: string;
  apiType: ApiType;
  phase: ExecutionPhase;
  steps?: StepV4[];
  deployedAt?: Date;
  createdAt?: Date;
  updatedAt?: Date;
}
